### ISSUE TITLE
Do metric internals need to be batched?

### ISSUE THREAD

**Issue by jack89roberts** [alan-turing-institute/ARC-LoCoMoSeT#30]
Currently, both PARC and Renggli take the whole dataset (subset) as input and compute the metric in one go. With PARC, for example, this could mean creating a big matrix which can use up RAM pretty rapidly if the data/feature dimension becomes large.

Consider whether we need the PARC and Renggli implementations to be able to work in batches (and how much work it would be to make that possible).

**Comment by eddableheath** [comment]
Did we get to a decision with this? 

- We're dropping PARC, however whilst the size of the metric grows as the square of the number of datapoints it doesn't need a huge probe set. 
- For LogME the size of the array doesn't increase from the input so what ever the probe set used is what it's capped at. I wouldn't have thought it'd become and issue, however batching LogME would require adapting the implementation we're using. The other issue i could see with LogME is that as a method it requires fitting the data, i.e. ideally we'd need to the entire probe set at once to get a reasonable score.
- I would have thought that Renggli we'd have a similar issue to the second LogME issue, i.e. Renggli uses a fitting method which I wouldn't have thought could be batched?

Even if the fitting methods being batched makes sense it might be quite a lot of refactoring to batch the metrics, however currently I don't think this is necessary.

**Comment by jack89roberts** [comment]
We can leave it until we have a case where it's definitely a problem I think. It might come up if we either use a dataset where we need to use more images than the ImageNet-1k validation set, or if we use a model that has a much bigger output dimension (more features) than the ones we've considered so far. But otherwise we should be ok.

For Renggli we could fit a single-layer neural net on top of the features to make it batched, which shouldn't be much work in terms of code but there'd then be a lot more to worry about in terms of hyperparameters/ensuring it fits correctly. For the other metrics, I'm less clear on how much work it would be to batch them (and whether that would even be possible).

**Comment by eddableheath** [comment]
I don't think that LogME, NCE or LEEP could be batched, just given the properties of the mathematical functions for each metric, i.e. they're not linear and don't have any linear subroutines/functions such that we can have f(D) = f(d_1) + f(d_2) (or similar combinations) for batches d_i of the dataset D.

**Comment by eddableheath** [comment]
You could batch PARC, however I think this would degrade the quality of the metric (not that we've seen it be particularly high quality anyway...) as this would be the equivalent to computing distance matrix of the form D = direct_sum_i(d_i) for batched distance matrices d_i, i.e. a block diagonal form where each block is a distance matrix for the batch. This would still function, however it would be large loss of information.

**Comment by jack89roberts** [comment]
Maybe we should have an issue about tracking how both time taken and RAM used scales with input size for all the metrics as a precursor to/instead of this. We have some times saved already but only looking at no. samples, not no. features, and we haven't done anything with RAM. We could also look at using FLOPs rather than time.
